### PR TITLE
[WIP] Lua Plugin for Handwriting recognition

### PR DESCRIPTION
**Pull request by rolandlo**
This is a Lua Plugin that uses the Google IME API for handwriting recognition.

The current initial version only supports recognition of the handwritten text on the selected layer, although the UI shows more options. The language and the number of returns (i.e. the number of alternatives for the recognized text, that are displayed) already work.

This Plugin requires the following lua-modules 
- `lgi `(for creating the UI)
- `socket`,` ssl.https` and `ltn12` (for the `https-request`)

These modules, which can be installed with the `luarocks` package manager, will be searched for in` /usr/local/sahre/lua/5.3/`
If that is the wrong path, the `initUi `function of `recognition.lua` must be modified.

### CODE REVIEW

**Comment by rolandlo** [comment]
Current progress can be seen from this video: 
textRecognitionDemo
Basically everything works, just some fine-tuning left.

**Comment by rolandlo** [comment]
As it turns out, the google server is only able to handle at most 150 strokes (roughly 2 handwritten sentences) at the time. That is quite unfortunate. Splitting up the strokes into chunks of roughly 150 strokes is possible, but it is not trivial to patch the recognized text pieces together.

**Comment by amirji** [comment]
I was thinking about this API:
https://developer.myscript.com/ 
Much more supported than google API.

**Comment by rolandlo** [comment]
@amirji I have seen the myscript API. It certainly looks pretty powerful. But only a limited number of requests per month are free. After that it quickly gets expensive. Moreover you need a subscription and only iOS, Android, Windows and Web are supported; Linux support is missing. That makes it problematic.

**Comment by rolandlo** [comment]
This will stall for a while until I have extended the Lua Plugin API. It is better to do that in a consistent manner than to extend the API time by time just to make one single plugin work.

**Comment by DeveloperOl** [comment]
Why are you using the google online recognition and not something like _libtesseract_ from https://github.com/tesseract-ocr/tesseract ?

**Comment by rolandlo** [comment]
@DeveloperOl See #2152 for the reason. In short: Tesseract is off-line text recognition.

**Comment by rolandlo** [comment]
An interesting alternative to the Google server would be [MathPix](https://docs.mathpix.com/). It has a documented API and not only returns handwritten text, but also math. So you potentially could use the selection tool in Xournal++ to select a handwritten formula, activate the plugin via menu or via keyboard accelerator, get the formula translated to LaTeX by Mathpix and Xournal++ would paste it as an editable LaTeX object. Similarly with handwritten text (or even a combination of both).

**Comment by abrac** [comment]
*Just brainstorming*: I found this cool open-source program called [CellWriter](https://github.com/risujin/cellwriter). It is able to do character recognition based on input from a mouse/pen. I wonder if it is possible to pipe out handwritten strokes from Xournal++ to CellWriter... 🤔

**Comment by rolandlo** [comment]
@abrac The problem is, that CellWriter expects the letters to be written in cells, whereas in Xournal++ you can write freely across the whole document.

**Comment by abrac** [comment]
That's true! Maybe the code from CellWriter could still be adapted to somehow work with Xournal++ 🤷. But it would probably take a whole PhD on its own to implement it that way! 🤣

**Comment by yingshaoxo** [comment]
Hey, guys!

How about we do not recognize the text.

Instead, we recognize math symbols, which are more important than pure text:


https://mathpix.com/

https://github.com/ritheshkumar95/im2latex-tensorflow

**Comment by rolandlo** [comment]
@yingshaoxo 
Yes, mathpix is an interesting option (see [above](https://github.com/xournalpp/xournalpp/pull/2176#issuecomment-740031495))
im2latex looks like a weaker solution, which would probably not work with handwritten formulas.

**Comment by amirji** [comment]
@yingshaoxo @rolandlo 
I am using mathpix daily and it is great but it is not free. Only the first 100 tries in each month are free. I am curious if we could make one by ourselves or not, the trained data can be downloaded and no need for a remote API.

**Comment by yingshaoxo** [comment]
We could make an additional local api server by making a python package to
PyPI, then any users who want to use it could just install a python package
to use it. (this step is easy, I'm an Python export.

But the problem is, I can not understand the logic of tensorflow, I could
only understand keras....such a pain...

**Comment by yingshaoxo** [comment]
I found an app: https://math.microsoft.com/en

One it recognized an image, you long press that image, you'll get latex copied into your clipboard.

And this app does not require login, we may use it to recognize.(if anyone knows how to capture network data packages, we could simulate that request to microsoft servers)

**Comment by personalizedrefrigerator** [comment]
@rolando Please break the changes to the plugin API into a separate PR (if you have time :)). It would be nice to have the additional stroke functionality for things like #2850.

**Comment by rolandlo** [comment]
> @rolando Please break the changes to the plugin API into a separate PR (if you have time :)). It would be nice to have the additional stroke functionality for things like #2850.

Not sure how `applib_getStrokes` (which returns a list of all strokes from a Xournal++ document in a Lua table) would help for #2850. In that issue the goal is to put strokes (obtained from some image) into the document, which is the other way round.

It wouldn't be much work to break the changes to the plugin API into a separate PR, if there is need for it. However there are still some things I am unsure about and which may become more difficult to change later, if there are already some plugins based on it:
- Currently a stroke in the Lua table consists of a table for the x-coordinates of the points and a table for the y-coordinates of the points. That's how the Google server (and other recognition software) want the strokes to be passed. The pressure data is ignored (which I think shouldn't). A stroke could also consist of a table of points, where each point is a table consisting of an x-, y- (and pressure-)value. I am not sure what is better.
- `applib_getStrokes` should eventually also yield the width of each stroke, the line style, the color, the fill opacity and the stroke tool (pen, highlighter or eraser) used. This doesn't need to be implemented all at once, but at least in a way, that doesn't require plugin authors to update their plugin each time the API is extended. 
-  Eventually there should also be `applib_getText`, `applib_getImage`, `applib_getTeXImage`
 - Eventually it should be possible to create and filter selections from the plugin API, to modify elements of a selection and to add elements to a selection. That's what will be needed for #2850. All in all this will require a decent amount of work.

**Comment by personalizedrefrigerator** [comment]
It looks like [QtVirtualKeyboard](https://doc.qt.io/qt-5/handwriting.html) uses [Lipi toolkit](http://lipitk.sourceforge.net/lipi-toolkit.htm) for its handwriting recognition... Maybe we could use this...

**Comment by WillNilges** [comment]
Hello, I'm wondering if y'all are still working on this? I'm starting to pick back up on working on this and would love to help out. (Hi it's the guy from #2850)

**Comment by rolandlo** [comment]
@WillNilges Hi! No, I am not working on this currently and I have no plan to do so in the near future. You are welcome to work on this.

**Comment by kauesena** [comment]
That would be very nice as then files would be searchable!

**Comment by IloveJesusmyKing** [comment]
Has anyone looked into the recognition technology that Microsoft uses for OneNote. I have found that their setup is efficient in processing lots of handwriting. It allows for searching handwriting.

**Comment by All3xJ** [comment]
Any news about this amazing plugin?

**Comment by yingshaoxo** [comment]
Don't you guys ever think about using Python?

https://github.com/JaidedAI/EasyOCR

**Comment by WillNilges** [comment]
I've thought about it, and personally, I'd really like it if we could figure out a way to do completely local recognition that didn't rely on proprietary software. This could be a viable option. And I'm looking at integrating python OpenCV into https://github.com/willnilges/inkpath so hopefully I'll have some relevant experience doing that soon.

**Comment by amirji** [comment]
We can use one of these engines for local recognition:

https://github.com/kba/awesome-ocr

However, still, local math recognition will be a challenge.

**Comment by Zireael07** [comment]
I think math support could be done as a follow-up thing?

**Comment by rolandlo** [comment]
Here is a [link](https://wicg.github.io/handwriting-recognition/) to a suggested Handwriting Recognition API that seems to fit well with what we would love to achieve in Xournal++. It includes segmentation information about each transcription, which means that the strokes in Xournal++ could be matched with text characters.

File: plugins/HandwritingRecognition/plugin.ini [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[about]
+## Author / Copyright notice
+author=Roland Lötscher
+
+description=Issue https://github.com/xournalpp/xournalpp/issues/2152
+
+## If the plugin is packed with Xournal++, use
+## <xournalpp> then it gets the same version number
+version=<xournalpp>
```
Comment (lucgiffon) [inline plugins/HandwritingRecognition/plugin.ini:9]: Hello! I don't understand. We should replace <xournallpp> with the actual version number?

**Comment by Captain-Pumpkinhead** [comment]
Any updates?  I'd love to use this plugin.  It'd make integrating Xournal++ with Obsidian much easier.

**Comment by michucz** [comment]
Can't wait to test this once it is available. One question, are you planning, or is it even possible to male the handwritten text searhable after it has been recognized? That would be amazing.

**Comment by rolandlo** [comment]
Please check out https://github.com/PellelNitram/xournalpp_htr for a more promising approach than this plugin.

**Comment by rolandlo** [comment]
Closing in favor of https://github.com/PellelNitram/xournalpp_htr

**Pull request closed without merging**